File: Xamarin/Xamarin.Native/Xamarin.Native.iOS/ViewControllers/LocationTestController.cs
using CoreFoundation;
using Foundation;
using System;
using System.Diagnostics;
using System.Drawing;
using UIKit;
using Xamarin.iOS;

namespace Xamarin.Native.iOS.ViewControllers
{
    public partial class LocationTestController : UIViewController
    {
        Stopwatch stopwatch;
        LocationTestService locationService;

        public LocationTestController(IntPtr handle) : base(handle)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            locationService = new LocationTestService();
            locationService.LocationChanged += LocationChanged;
        }

        partial void StartPositioning(UIButton sender)
        {
            RefreshUI(true);

            stopwatch = new Stopwatch();
            stopwatch.Start();

            try
            {
                locationService.GetLocation();
            }
            catch (LocationUnavailableException)
            {
                DispatchQueue.MainQueue.DispatchAsync(() =>
                {
                    UIAlertView alert = new UIAlertView() { Title = "Błąd", Message = "Lokalizacja niedostępna" };
                    alert.AddButton("OK");
                    alert.Show();
                });
            }
        }

        private void LocationChanged(double latitude, double longitude)
        {
            stopwatch.Stop();
            DispatchQueue.MainQueue.DispatchAsync(() =>
            {
                positionLabel.Text = string.Format("Długość: {0}\nSzerokość: {1}", Math.Round(longitude, 4), Math.Round(latitude, 4));
                timeLabel.Text = stopwatch.GetDurationInSeconds();
                RefreshUI(false);
            });
        }

        private void RefreshUI(bool isPositioning)
        {
            startButton.Hidden = isPositioning;
            activityIndicator.Hidden = !isPositioning;
        }
    }
}

File: Native.iOS/Native.iOS/ViewControllers/LocationTestController.m
//
//  LocationTestController.m
//  Native.iOS
//
//  Created by Pawel Grzmil on 08/05/16.
//  Copyright © 2016 Pawel Grzmil. All rights reserved.
//

#import "LocationTestController.h"
#import "LocationTestService.h"
#import "Stopwatch.h"


@interface LocationTestController () <LocationTestServiceDelegate>

@property (weak, nonatomic) IBOutlet UIButton *startButton;
@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (weak, nonatomic) IBOutlet UILabel *positionLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;

@property (strong, nonatomic) Stopwatch *stopwatch;
@property (strong, nonatomic) LocationTestService *locationService;

@end

@implementation LocationTestController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Test pozycji GPS";
    
    self.stopwatch = [Stopwatch new];
    self.locationService = [LocationTestService new];
    self.locationService.delegate = self;
}

- (IBAction)startPositioning:(id)sender {
    [self refreshUI:true];
    [self.stopwatch start];
    
    [self.locationService getLocation];
}

- (void)locationChanged:(double)latitude longitude:(double)longitude{
    [self.stopwatch stop];
    [self.locationService stop];
    dispatch_async( dispatch_get_main_queue(), ^{
        self.positionLabel.text = [NSString stringWithFormat:@"Długość: %f\nSzerokość: %f", latitude, longitude];
        self.timeLabel.text = [NSString stringWithFormat:@"%@\n%@", [self.stopwatch getDurationInSeconds], [self.stopwatch getDurationInMilliseconds]];
        [self refreshUI:false];
    });
}

- (void)refreshUI:(BOOL)isPositioning{
    self.activityIndicator.hidden = !isPositioning;
    self.startButton.hidden = isPositioning;
}
@end
